linera-indexer/macro/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+}
+
+fn wrap_register(reg: &mut ImplItemFn) {
+    let mut block = reg.block.clone();
+    SelfReplace.visit_block_mut(&mut block);
+    let stmts: &mut Vec<Stmt> = &mut block.stmts;
+    let stmt: Stmt = parse_quote! {
+        let mut _plugin = self.0.lock().await;
+    };
+    let mut stmts_new = vec![stmt];
+    stmts_new.append(stmts);
+
+    if let Some(stmt) = stmts_new.pop() {
+        match stmt.clone() {
+            Stmt::Expr(Expr::Call(call), None) => match *call.func.clone() {
+                Expr::Path(path) => {
+                    if &path.path.segments[0].ident.to_string() == "Ok" {
+                        let expr = call.args[0].clone();
+                        let stmt: Stmt = parse_quote! {
+                            #expr;
+                        };
+                        stmts_new.push(stmt)
+                    } else {
+                        stmts_new.push(stmt)
+                    }
+                }
+                _ => stmts_new.push(stmt),
+            },
+            _ => stmts_new.push(stmt),
+        }
+    }
+    let stmt: Stmt = parse_quote! {
+        return Ok(_plugin.save().await?);
+    };
+    stmts_new.push(stmt);
+    reg.block.stmts = stmts_new
+}
+
+fn find_register(items: &[ImplItem]) -> ImplItemFn {
+    let mut register_function = items
+        .iter()
+        .find_map(|item| {
+            let ImplItem::Fn(fun_item) = item else {
+                return None;
+            };
+            if &fun_item.sig.ident.to_string() == "register" {
+                Some(fun_item.clone())
+            } else {
+                None
+            }
+        })
+        .expect("register function not found");
+    wrap_register(&mut register_function);
+    register_function
+}
+
+fn public_functions(items: &[ImplItem]) -> Vec<ImplItem> {
+    items
+        .iter()
+        .filter_map(|item| {
+            let ImplItem::Fn(mut fun_item) = item.clone() else {
+                return None;
+            };
+            match (
+                fun_item.sig.ident.to_string().as_str(),
+                fun_item.vis.clone(),
+            ) {
+                ("register", _) => None,
+                (_, Visibility::Public(_)) => {
+                    wrap_fn(&mut fun_item);
+                    Some(ImplItem::Fn(fun_item))
+                }
+                _ => None,
+            }
+        })
+        .collect()
+}
+
+fn private_functions(items: &[ImplItem]) -> Vec<ImplItem> {
+    items
+        .iter()
+        .filter_map(|item| {
+            let ImplItem::Fn(fun_item) = item.clone() else {
+                return None;
+            };
+            match (
+                fun_item.sig.ident.to_string().as_str(),
+                fun_item.vis.clone(),
+            ) {
+                ("register", _) => None,
+                (_, Visibility::Inherited) => Some(ImplItem::Fn(fun_item)),
+                _ => None,
+            }
+        })
+        .collect()
+}
+
+fn plugin_impl(
```

Comment:
Apart from `static_name`, which would have to be manual then, it seems this could be a `struct<P> PluginWrapper(Arc<Mutex<P>>)` instead of a macro, with something like:

```rust
impl<DB, P> Plugin<DB> for PluginWrapper<P>
    where P: InnerPlugin, DB: …
```

That could also call `register`, so `find_register` and `wrap_register` wouldn't be needed.

Reply:
I will try to work something out to have less things implemented in the macro

Reply:
I'm stuck with the issue that if I use a wrapper struct, I cannot implement `async_graphql::ObjectType` on it

Reply:
One idea would be to `impl<T, P> Plugin for T where T: Deref<T>, P: Plugin`.

Then one could `impl Deref` for a `struct ActualPluginWrapper(PluginWrapper<MyPlugin>)`, and derive the `async_graphql` stuff for `ActualPluginWrapper`.

Maybe `PluginWrapper` could have a convenience method for locking. Then we wouldn't need a macro at all?

Reply:
I kind of understand, but I am not familiar with `Deref` and I am not managing to make it work.